v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/handler.ts

```diff
@@ -0,0 +1,294 @@
+    }
+
+    if (rpcReq !== undefined) {
+      rpcResp.id = rpcReq.id !== undefined ? rpcReq.id : null;
+    }
+
+    return rpcResp;
+  }
+
+  async #handleSingleWsRequest(
+    rpcReq: JsonRpcRequest,
+    subscriptions: string[],
+  ) {
+    const rpcResp = await this.#handleSingleRequest(rpcReq);
+
+    // If eth_subscribe was successful, keep track of the subscription id,
+    // so we can cleanup on websocket close.
+    if (
+      rpcReq.method === "eth_subscribe" &&
+      isSuccessfulJsonRpcResponse(rpcResp) &&
+      typeof rpcResp.result === "string"
+    ) {
+      subscriptions.push(rpcResp.result);
+    }
+
+    return rpcResp;
+  }
+
+  readonly #handleRequest = async (
+    req: JsonRpcRequest,
+  ): Promise<JsonRpcResponse> => {
+    const result = await this.#provider.request({
+      method: req.method,
+      params: req.params,
+    });
+
+    return {
+      jsonrpc: "2.0",
+      id: req.id,
+      result,
+    };
+  };
+}
+
+const _readJsonHttpRequest = async (req: IncomingMessage): Promise<any> => {
+  let json;
+
+  try {
+    const buf = await getRawBody(req);
+    const text = buf.toString();
+
+    json = JSON.parse(text);
+  } catch (error) {
+    if (error instanceof Error) {
+      // eslint-disable-next-line no-restricted-syntax -- Malformed JSON-RPC request received, report to user as a json input error.
+      throw new InvalidJsonInputError(`Parse error: ${error.message}`);
+    }
+
+    throw error;
+  }
+
+  return json;
+};
+
+const _readWsRequest = (msg: string): JsonRpcRequest | JsonRpcRequest[] => {
+  let json: any;
+  try {
+    json = JSON.parse(msg);
+  } catch (error) {
+    if (error instanceof Error) {
+      // eslint-disable-next-line no-restricted-syntax -- Malformed JSON-RPC request received, report to user as a json input error.
+      throw new InvalidJsonInputError(`Parse error: ${error.message}`);
+    }
+    throw error;
+  }
+
+  return json;
+};
+
+const _handleError = (error: any): JsonRpcResponse => {
+  // extract the relevant fields from the error before wrapping it
+  let txHash: string | undefined;
+  let returnData: string | undefined;
+
+  if (error.transactionHash !== undefined) {
+    txHash = error.transactionHash;
+  }
+  if (error.data !== undefined) {
+    if (error.data?.data !== undefined) {
+      returnData = error.data.data;
+    } else {
+      returnData = error.data;
+    }
+
+    if (txHash === undefined && error.data?.transactionHash !== undefined) {
+      txHash = error.data.transactionHash;
+    }
+  }
+
+  // In case of non-hardhat error, treat it as internal and associate the appropriate error code.
+  if (!ProviderError.isProviderError(error)) {
+    error = new InternalError(error);
+  }
+
+  const response: FailedJsonRpcResponse = {
+    jsonrpc: "2.0",
+    id: null,
+    error: {
+      code: error.code,
+      message: error.message,
+    },
+  };
+
+  const data: any = {
+    message: error.message,
+  };
+
+  if (txHash !== undefined) {
```

Comment:
why is this block empty?

Reply:
A great catch, I have no idea how that happened.